src/Http/Routing/src/Matching/HeaderMatcherPolicy.cs

```diff
@@ -0,0 +1,207 @@
+
+namespace Microsoft.AspNetCore.Routing
+{
+    internal sealed class HeaderMatcherPolicy : MatcherPolicy, IEndpointComparerPolicy, IEndpointSelectorPolicy
+    {
+        /// <inheritdoc/>
+        // Run after HttpMethodMatcherPolicy (-1000) and HostMatcherPolicy (-100), but before default (0)
+        public override int Order => -50;
+
+        /// <inheritdoc/>
+        public IComparer<Endpoint> Comparer => new HeaderMetadataEndpointComparer();
+
+        /// <inheritdoc/>
+        bool IEndpointSelectorPolicy.AppliesToEndpoints(IReadOnlyList<Endpoint> endpoints)
+        {
+            _ = endpoints ?? throw new ArgumentNullException(nameof(endpoints));
+
+            // When the node contains dynamic endpoints we can't make any assumptions.
+            if (ContainsDynamicEndpoints(endpoints))
+            {
+                return true;
+            }
+
+            return AppliesToEndpointsCore(endpoints);
+        }
+
+        /// <inheritdoc/>
+        public Task ApplyAsync(HttpContext httpContext, CandidateSet candidates)
+        {
+            _ = httpContext ?? throw new ArgumentNullException(nameof(httpContext));
+            _ = candidates ?? throw new ArgumentNullException(nameof(candidates));
+
+            for (int i = 0; i < candidates.Count; i++)
+            {
+                if (!candidates.IsValidCandidate(i))
+                {
+                    continue;
+                }
+
+                var metadata = candidates[i].Endpoint.Metadata.GetMetadata<IHeaderMetadata>();
+                var metadataHeaderName = metadata?.HeaderName;
+                if (string.IsNullOrEmpty(metadataHeaderName))
+                {
+                    // Can match any request
+                    continue;
+                }
+
+                var metadataHeaderValues = metadata.HeaderValues;
+                int maxValuesToInspect = metadata.MaximumValuesToInspect;
+
+                var matched = false;
+                if (httpContext.Request.Headers.TryGetValue(metadataHeaderName, out var requestHeaderValues))
+                {
+                    if (metadataHeaderValues?.Count == 0)
+                    {
+                        // We were asked to match as long as the header exists, and it *does* exist
+                        matched = true;
+                    }
+                    else
+                    {
+                        var comparisonFunc = GetComparisonFunc(metadata.ValueMatchMode);
+                        var stringComparison = metadata.ValueIgnoresCase ? StringComparison.OrdinalIgnoreCase : StringComparison.Ordinal;
+                        for (int j = 0; j < metadataHeaderValues.Count; j++)
+                        {
+                            for (int k = 0; k < requestHeaderValues.Count && k < maxValuesToInspect; k++)
+                            {
+                                if (comparisonFunc(requestHeaderValues[k], metadataHeaderValues[j], stringComparison))
```

Comment:
Can we store these on local variables? It will avoid indexing the collections repeatedly.